source/deprecation-policy.rst

```diff
@@ -0,0 +1,145 @@
+---
+layout: default
+title: Deprecation Policy
+menuSlug: deprecation-policy
+permalink: /deprecation-policy/index.html
+---
+
+Deprecation Policy
+==================
+
+The Doctrine team will adhere to this deprecation policy whenever it is reasonably possible.
+
+Using ``@deprecated`` and ``PHPStan Deprecation Rules``
+-------------------------------------------------------
+
+The primary and preferred method for deprecating functionality within Doctrine is by using
+the ``@deprecated`` annotation. To have usages of functionality in your dependencies analyzed
+and reported, you must use a tool like `PHPStan Deprecation Rules`_ or `Pslam Static Analysis`_.
+
+Here is an example of how this would be used in Doctrine:
+
+.. code-block:: php
+
+    namespace Doctrine\ORM;
+
+    class EntityManager
+    {
+        // ...
+
+        /**
+         * @deprecated
+         */
+        public function getSomething() : int
+        {
+            // ...
+        }
+    }
+
+Then, if your code directly depends on ``doctrine/orm`` which provides the
+``Doctrine\ORM\EntityManager`` class, you will get a warning about it when you run
+`PHPStan Deprecation Rules`_ or `Pslam Static Analysis`_. Here is an example:
+
+.. code-block:: php
+
+    namespace App;
+
+    use App\Entities\User;
+    use Doctrine\ORM\EntityManager;
+
+    class UserRegistration
+    {
+        /** @var EntityManager */
+        private $entityManager;
+
+        // ...
+
+        public function create(array $userData) : User
+        {
+            $something = $this->entityManager->getSomething();
+
+            // ...
+        }
+    }
+
+Using ``trigger_error()`` and ``phpunit-bridge``
+------------------------------------------------
+
+In some cases, you may need to conditionally deprecate functionality. In these cases, it is required
+to use the ``trigger_error()`` function and a tool like `PHPUnit Bridge`_ which implements a custom
```

Comment:
Discussed this with @nicolas-grekas in person: the `PHPUnit_Bridge` is not needed. With out of the box PHPUnit, this is the behavior:

Passes:

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

final class SomeTest extends \PHPUnit\Framework\TestCase
{
    public function testFoo() : void
    {
        @trigger_error(E_USER_DEPRECATED, 'potato');
        self::assertTrue(true);
    }
}
```
```sh
./vendor/bin/phpunit SomeTest.php 
PHPUnit 7.3.2 by Sebastian Bergmann and contributors.

.                                                                   1 / 1 (100%)

Time: 60 ms, Memory: 4.00MB

OK (1 test, 1 assertion)
```


Fails:

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

final class SomeTest extends \PHPUnit\Framework\TestCase
{
    public function testFoo() : void
    {
        trigger_error(E_USER_DEPRECATED, 'potato');
        self::assertTrue(true);
    }
}
```
```sh
./vendor/bin/phpunit SomeTest.php 
PHPUnit 7.3.2 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 60 ms, Memory: 4.00MB

There was 1 error:

1) SomeTest::testFoo
trigger_error() expects parameter 2 to be integer, string given

/home/ocramius/foo/SomeTest.php:9

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

Note that this is because PHPUnit by default checks the error level before turning it into an error, while following still leads to failures:

```php
<?php

set_error_handler(function () {
    echo 'I died :-(';
    die;
});

@trigger_error(E_USER_DEPRECATED, 'foo');
```

Produces:

```
I died :-(
```

This means that users that just use `set_error_handler()` (no parameters, no filtering, etc) will get a BC break.

As per discussion with @nicolas-grekas (where this surfaced as well), this is a decision to be taken, because some users may be doing it "wrong", but effectively I've already seen codebases where zero-warnings-or-notices-ever are accepted, and those codebases will be damaged by using `@trigger_error()`.

In my personal opinion, we should keep things towards `@deprecated`.

Reply:
@Ocramius I think you inverted the order of arguments in `trigger_error() ` calls.

> In my personal opinion, we should keep things towards `@deprecated`. 

I agree, we should strive for using `@deprecated` as much as possible.

Reply:
Indeed, the arguments were swapped :-\